### ISSUE TITLE
[bug] Wrong conan graph build-order in package_revision_mode

### ISSUE THREAD

**Issue by radonish** [conan-io/conan#6232]
### Environment Details
  * Operating System: Red Hat Enterprise Linux 7.4
  * Compiler: GCC 4.8.5
  * Conan version: 1.21.0
  * Python version: 3.7.1

### Steps to reproduce
Assumptions:
- Cache is clean; no binary packages exist so the expectation is a full build on the CI server will be necessary.
- libA depends on libC
- libB has no dependencies
- libC has no dependencies
- appX depends on libA and libB

Steps:
1. conan export libA ...
2. conan export libB ...
3. conan export libC ...
4. conan export appX ...
5. conan graph lock appX --lockfile appX.lock
6. conan build-order --build missing appX.lock

Package libA shows up as Unknown
Package libB shows up as Missing
Package libC shows up as Missing
Package appX shows up as Unknown

The problem is that the conan graph build-order output excludes libA since it is "Unknown". I expect that of appX, I suppose, since it's the application whose dependencies I'm trying to determine need building - but not of libA.

Thanks

**Comment by memsharded** [comment]
Hi @radonish 

Sorry for the delay in responding this.
I have been having a look at this, first a quick tip: at the moment you should also use ``conan graph lock appX --build=missing`` in the command ``graph lock`` too. Otherwise the lock might not work as expected.

Then, the behavior is correct. It is doing exactly what it is expected to do. "Unknown ID" nodes are not in the list of "to build" nodes because we don't know yet if they need to be built or not. It is possible that they don't. With ``package_revision_mode``, the dependencies resolution is exact, if after evaluating the package ID it is found as binary, then really no need to re-build it.

The way the package_revision_mode and build-order is intended to work is incremental, which by the way can be also more resource efficient:

- Compute the build-order
- Take the first level, fire those in parallel
- The moment any of those parallel jobs finish and some computing resources are available, compute the updated lock and **recompute** the build-order. This recomputation will also compute the final package ID that will be used instead of the Unknown one.
- Take that build order and repeat, making sure not relaunching jobs that are already running.

In this way, the utilization of resources is more effective. Imagine that the first level has 10 packages, 9 header-only libraries and one very heavy to build static library. Until the 10 finish, more builds cannot be launch, so 9 jobs will be not used until the static library finish. Packages depending on the header-only libraries could start as soon as those finish.

Said that, I think that it might be possible to add those to the output of build-order (without the package ID). I'll give it a try.

**Comment by memsharded** [comment]
This would be a possible approach: https://github.com/conan-io/conan/pull/6251

**Comment by radonish** [comment]
@memsharded, thanks for the detailed response. The approach you present certainly makes sense. The key thing we did not realize was that the intended flow was to run build-order incrementally.

I wrote a quick shell script to create the structure above to see the process in a simpler/quicker way than our product baselines (see the attached shell script). I ran the conan commands manually at the bottom of the script but included them within for reference; I see it working exactly as you describe.

Thanks

[create_mock_projects.sh.txt](https://github.com/conan-io/conan/files/3980539/create_mock_projects.sh.txt)

**Comment by memsharded** [comment]
Cool, thanks for following up and reporting

> The key thing we did not realize was that the intended flow was to run build-order incrementally.

We have in our TODO list to document this better and add more examples, to make this more evident. Hopefully soon.

> I wrote a quick shell script to create the structure above to see the process in a simpler/quicker way than our product baselines (see the attached shell script). I ran the conan commands manually at the bottom of the script but included them within for reference; I see it working exactly as you describe.

Certainly the best. I always write small, scripted proof of concepts of flows before doing that in production or CI, which is a nightmare to debug. Thanks for sharing your script!

Leaving this issue open at the moment, because we might still add the UnknownID to the output of build-order, to be discussed based on the PR.

**Comment by radonish** [comment]
@memsharded @lasote, related to this issue and using `conan graph` to determine the build order for CI tools, is it possible to get some updated blog posts/examples now that the old method of using `conan info` for build order is deprecated?

I'm referring to these resources:
- https://jenkins.io/blog/2017/07/07/jenkins-conan/
- https://github.com/lasote/skynet_example/blob/master/multi_build.groovy
- https://github.com/lasote/skynet_example/blob/master/single_build.groovy

I'm attempting to get things working with Jenkins based on these examples; I'll create a separate issue after I get deeper in, but right now I'm working through what appears to be formatting issues with the build-order output. Perhaps something changed with the migration from `conan info --build-order` to `conan graph lock` + `conan graph --build-order`.

Thanks!

**Comment by memsharded** [comment]
Lets try to get this into next 1.22.

Yes, updating the docs and examples to include exactly these things, is a high priority for us, but unfortunately, many other small things pile up, keeping us away of doing it :( We will eventually find some time and do it, thanks for the feedback!

**Comment by memsharded** [comment]
Merged. Will be released in 1.22.
